rgw: add custom endpoint list option for zone (#10629)
rook/rook: Documentation/CRDs/Object-Storage/ceph-object-multisite-crd.md

    erasureCoded:
      dataChunks: 2
      codingChunks: 1

>   customEndpoints:

Reviewer: Looks like there is an extra space so it looks like it's a setting under `dataPool`, but it's a top level setting, right?

Author: Yes it top level setting, in my vim editor it is showing as same level as `dataPool`, not sure why it is list like that. Can you please give it as suggestion, so that I can directly commit to this PR

Reviewer: It should be using spaces. Do you have a tab instead? If it's already spaces, just remove one space.

* `zonegroup`: The object zonegroup in which the zone will be created. This matches the name of the object zone group CRD.
* `metadataPool`: The settings used to create all of the object store metadata pools. Must use replication.
* `dataPool`: The settings to create the object store data pool. Can use replication or erasure coding.

> * `customEndpointList`: If this zone cannot be accessed from other peer Ceph clusters via the ClusterIP Service endpoint created by Rook, you must set this to the externally reachable endpoint(s). You must also include the port in the definition. For example: "https://my-object-store.my-domain.net:443". In many cases, you should set this to the endpoint of the ingress resource that makes the CephObjectStore associated with this CephObjectStoreZone reachable to peer clusters. The list can have one or more endpoints pointing to different RGW servers and need to list all the endpoint(s) part of zone endpoint list in multisite configuration

Reviewer 2: I think a lot of users won't fully understand how a zone relates to a store. Let's be as clear as possible for them in the description for this field and how to use it.

```suggestion
* `customEndpointList`: Specify the endpoint(s) that will accept multisite replication traffic for this zone. You must also include the port in the definition. For example: "https://my-object-store.my-domain.net:443". By default, Rook will set this to the DNS name of the ClusterIP Service created for the CephObjectStore that corresponds to this zone. 

  Most multisite configurations will not exist within the same Kubernetes cluster, meaning the default value will not be useful. In these cases, you will be required to create your own custom ingress resource for the CephObjectStore in order to make the zone available for replication. You must add the endpoint for your custom ingress resource to this list to allow the store to accept replication traffic.

  In the case of multiple stores (or multiple endpoints for a single store), you are not required to put all endpoints in this list. Only specify the endpoints that should be used for replication traffic.
```

This should render as shown below this line. @thotz @alimaredia are the details I included here still correct?

-----

* `customEndpointList`: Specify the endpoint(s) that will accept multisite replication traffic for this zone. You must also include the port in the definition. For example: "https://my-object-store.my-domain.net:443". By default, Rook will set this to the DNS name of the ClusterIP Service created for the CephObjectStore that corresponds to this zone. 

  Most multisite configurations will not exist within the same Kubernetes cluster, meaning the default value will not be useful. In these cases, you will be required to create your own custom ingress resource for the CephObjectStore in order to make the zone available for replication. You must add the endpoint for your custom ingress resource to this list to allow the store to accept replication traffic.

  In the case of multiple stores (or multiple endpoints for a single store), you are not required to put all endpoints in this list. Only specify the endpoints that should be used for replication traffic.

Reviewer 2: We should also make a note about the operator needing restarted if this list is reset to empty.

Reviewer: If the list is updated, wouldn't a new reconcile be triggered? Or why is a full operator restart needed?

Author: @BlaineEXE : RGW won't allow the endpoint list to be empty, so even if Rook removes the last endpoint(s) from zone/zonegroup. RGW will still keep the last endpoint. During the restart of the operator pod after emptying the `customEndpoints` Rook will update the endpoint list with rgw services names with last endpoint(s) plus with service names of rgws. But the user still needs to remove the invalid endpoint manually. So IMO it is better not this approach.

@travisn : Restart of operator will triggers in the reconciliation of objectstores CR, so all the rgw endpoints added seeing `CustomEndpoints` in Zone is empty.

Reviewer 2: > But the user still needs to remove the invalid endpoint manually. So IMO it is better not this approach.

I'd like to figure out how to avoid the user having to do any manual modification unless necessary. If it is going to be exceedingly complicated, I think it's fine to have this as a documented corner case for our initial implementation, but we should document it.

* `zonegroup`: The object zonegroup in which the zone will be created. This matches the name of the object zone group CRD.
* `metadataPool`: The settings used to create all of the object store metadata pools. Must use replication.
* `dataPool`: The settings to create the object store data pool. Can use replication or erasure coding.

> * `customEndpoints`:  Specify the endpoint(s) that will accept multisite replication traffic for this zone. You must also include the port in the definition. For example: "https://my-object-store.my-domain.net:443". By default, Rook will set this to the DNS name of the ClusterIP Service created for the CephObjectStore that corresponds to this zone.

Reviewer 2: ```suggestion
* `customEndpoints`:  Specify the endpoint(s) that will accept multisite replication traffic for this zone. You may include the port in the definition if necessary. For example: "https://my-object-store.my-domain.net:443". By default, Rook will set this to the DNS name of the ClusterIP Service created for the CephObjectStore that corresponds to this zone.
```
